internal/storage/image.go

```diff
@@ -420,6 +438,200 @@ func (svc *imageService) imageStatus(systemContext *types.SystemContext, unstabl
+		return "", fmt.Errorf("pull local manifest digest: %w", err)
+	}
+	log.Debugf(ctx, "Local manifest digest for image %q is: %s", imageParsed, localManifestDigest)
+
+	canonicalRef, err := reference.WithDigest(reference.TrimNamed(imageParsed), localManifestDigest)
+	if err != nil {
+		return "", fmt.Errorf("create canonical reference: %w", err)
+	}
+
+	return canonicalRef.String(), nil
+}
+
+func (svc *imageService) pullLocalManifestDigest(ctx context.Context, sys *types.SystemContext, policyContext *signature.PolicyContext, imageParsed reference.Named) (digest.Digest, error) {
+	srcRef, err := docker.NewReference(imageParsed)
+	if err != nil {
+		return "", fmt.Errorf("creating docker:// reference for %q: %w", imageParsed.String(), err)
+	}
+
+	destRef, err := istorage.Transport.NewStoreReference(svc.store, imageParsed, "")
+	if err != nil {
+		return "", fmt.Errorf("creating containers-storage: reference for %q: %w", imageParsed.String(), err)
+	}
+
+	manifestBytes, err := copy.Image(ctx, policyContext, destRef, srcRef, &copy.Options{
+		SourceCtx:      sys,
+		DestinationCtx: sys,
+	})
+	if err != nil {
+		var e errcode.Error
+		if as := errors.As(err, &e); as && e.ErrorCode().Descriptor().HTTPStatusCode == http.StatusUnauthorized {
+			// We cannot support private registries right now, but may with future support from the CRI.
+			log.Errorf(ctx, "Verifying policies against private registries it not supported yet: %v", err)
+			return "", nil
+		}
+
+		var policyErr signature.PolicyRequirementError
+		if errors.As(err, &policyErr) {
+			return "", fmt.Errorf("%w: %w", crierrors.ErrSignatureValidationFailed, err)
+		}
+
+		return "", fmt.Errorf("pulling %q to %q: %w", transports.ImageName(srcRef), transports.ImageName(destRef), err)
+	}
+	return digest.FromBytes(manifestBytes), nil
+}
+
+func (svc *imageService) IsRunningImageAllowed(ctx context.Context, systemContext *types.SystemContext, configImage, userSpecifiedImage string) error {
+	policy, err := signature.DefaultPolicy(systemContext)
+	if err != nil {
+		return fmt.Errorf("get default policy: %w", err)
+	}
+
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return fmt.Errorf("create policy context: %w", err)
+	}
+
+	defer policyContext.Destroy() //nolint:errcheck
+
+	var localManifestDigest digest.Digest
+	// On fresh pulls retrieve the full image and digest as configImage
+	if parsedConfigImage, err := reference.Parse(configImage); err == nil {
+		if canonical, isDigested := parsedConfigImage.(reference.Canonical); isDigested {
+			localManifestDigest = canonical.Digest()
+		}
+	}
+
+	var imageParsed reference.Named
+	// Check if the configImage is a local image ID if the image is already pulled
+	if localManifestDigest.String() == "" {
+		img, err := svc.GetStore().Image(configImage)
+		if err != nil && !errors.Is(err, storage.ErrImageUnknown) {
+			return fmt.Errorf("error retrieving image %s: %w", configImage, err)
+		}
+		if img != nil {
```

Comment:
perhaps double check the length of img here.

Reply:
@rphillips `img` is of Image type, https://github.com/cri-o/cri-o/blob/a934077689a3ff1dc09a4f4130de60dfabd74e88/vendor/github.com/containers/storage/images.go#L32

Hence the check for if it is `nil`.

Reply:
Right, but the length of the array could be zero, and we are indexing into it.

Reply:
@rphillips something like,
```
if img != nil && img.Names != nil && len(img.Names) > 0
```
?

Reply:
I'm going to retract my comment here... I think the better approach is to check if err != nil if the alternative case, on line 542, and remove the `if img != nil` entirely.

Reply:
you can have `err == nil` and `img == nil` at the same time. It just means that, there was no error looking up in the image store but the image doesn't exist in that store.